### PR TITLE
Support for server-initiated Shiny-to-OJS exporting

### PR DESCRIPTION
**Pull request by jcheng5**
This PR makes it possible for OJS code to depend on Shiny variables
that are not known by the client at the time that OJS is starting up.
Normally, if an OJS cell depends on a variable that is not known, you
get an "X is not defined" error; with this change, the OJS runtime
starts out in a lenient phase where undefined variables are treated as
pending promises. Only after the Shiny server function has sent its
first message does the lenient phase end, and undefined variables
become errors again.

### CODE REVIEW

**Comment by dragonstyle** [comment]
These test failures can be ignored - the failures were introduced by another commit. Working to fix the issue now.

File: src/resources/formats/html/ojs/ojs-bundle.js [hunk 1]
```diff
@@ -54,6 +54,10 @@ export class OJSInABox {
     // so any variables that are still missing do cause "x is not defined"
     // errors.
     this.pendingGlobals = {};
+    // When true, the mechanism described in the `this.pendingGlobals` comment
+    // is used. When false, the result of accessing undefined variables is just
+    // "x is not defined". This should be considered private, only settable via
+    // constructor or `killPendingGlobals`.
     this.allowPendingGlobals = allowPendingGlobals;
```
Comment (cscheid) [inline src/resources/formats/html/ojs/ojs-bundle.js:61]: I'm ok keeping this as-is, but I'm also wondering if we should be using the convention that underscore-prefixed variables are intended to be private.

File: src/resources/formats/html/ojs/ojs-bundle.js [hunk 2]
```diff
@@ -708,8 +731,18 @@ export function createRuntime() {
   });
   quartoOjsGlobal.obsInABox = obsInABox;
   if (isShiny) {
+    // When isShiny, OJSInABox is constructed with allowPendingGlobals:true.
+    // Our guess is that most shiny-to-ojs exports will have been defined
+    // by the time the server function finishes executing (i.e. session init
+    // completion); so we call `killPendingGlobals()` to show errors for
+    // variables that are still undefined.
```
Comment (cscheid) [inline src/resources/formats/html/ojs/ojs-bundle.js:738]: This is the point where we would handle things differently if we decide to use the `ojs-imports` YAML entry to explicitly encode the expected Shiny outputs, right?

**Comment by cscheid** [comment]
LGTM. I'm going to go ahead and merge. (The failing test is new but also unrelated. It has to do with puppeteer in the github actions workflow)

File: src/resources/formats/html/ojs/ojs-bundle.js (hunk 1 above)
Comment (jcheng5) [reply src/resources/formats/html/ojs/ojs-bundle.js:61]: I had the same instinct (also for the `globals` method).

File: src/resources/formats/html/ojs/ojs-bundle.js (hunk 2 above)
Comment (jcheng5) [reply src/resources/formats/html/ojs/ojs-bundle.js:746]: Yes, although I don't think there's any reason why we can't allow both simultaneously. (I haven't tested that)

**Pull request merged**